avro-fastserde/src/main/java/com/linkedin/avro/fastserde/PrimitiveFloatList.java

```diff
@@ -60,44 +76,63 @@ public PrimitiveFloatList(Collection<Float> c) {
    * @throws IOException on io errors
    */
   public static Object readPrimitiveFloatArray(Object old, Decoder in) throws IOException {
-    long l = in.readArrayStart();
-    if (l > 0) {
-      PrimitiveFloatList array = (PrimitiveFloatList) newPrimitiveFloatArray(old, (int) l);
+    long length = in.readArrayStart();
+    long totalLength = 0;
+
+    if (length > 0) {
+      PrimitiveFloatList array = (PrimitiveFloatList) newPrimitiveFloatArray(old);
+
       do {
-        for (long i = 0; i < l; i++) {
-          array.addPrimitive(in.readFloat());
-        }
-        l = in.arrayNext();
-      } while (l > 0);
+        // Allocate ByeBuffer of size array length * Float.BYTES to hold the array values
+        ByteBuffer byteBuffer = ByteBuffer.allocate((int)length * FLOAT_SIZE).order(ByteOrder.LITTLE_ENDIAN);
+        in.readFixed(byteBuffer.array(), 0, (int)(length * FLOAT_SIZE));
+        array.byteBuffers.add(byteBuffer);
+        totalLength += length;
+        length = in.arrayNext();
+      } while (length > 0);
+
+      setupElements(old, array, (int)totalLength);
       return array;
     } else {
-      return newPrimitiveFloatArray(old, 0);
+      return new PrimitiveFloatList(0);
     }
   }
 
+  private static void setupElements(Object old, PrimitiveFloatList list, int size) {
+    if (old instanceof PrimitiveFloatList) {
+      PrimitiveFloatList oldFloatList = (PrimitiveFloatList) old;
+      if (size <= oldFloatList.getCapacity()) {
+        // reuse the float array directly
+        oldFloatList.clear();
+      } else {
+        oldFloatList.resizeAndClear(size);
+      }
+      return;
+    }
+    list.elements = new float[size];
+    list.size = size;
```

Comment:
Looks like this is the only place where the `size` member variable is set, however, this code will not execute if the early termination above occurs, in which case, `size` would not be set... I think we should make sure to always set the `size` in the beginning, prior to deserializing any of the items.